File: 18.4sum.cs
using System;
using System.Collections.Generic;

public class Solution18 {
    public IList<IList<int>> FourSum(int[] nums, int target) {
        Array.Sort(nums);
        IList<IList<int>> result = new List<IList<int>>();
        for (int i = 0; i < nums.Length - 3; i++)
        {
            // this is to keep sure there's no duplicates
            if (i > 0 && nums[i] == nums[i - 1]) 
                continue;
            for (int j = i + 1; j < nums.Length - 2; j++)
            {
                //same logic for j, make sure there's no duplicates
                if (j > i + 1 && nums[j] == nums[j - 1]) 
                    continue;
                int sum = target - nums[i] - nums[j];
                int left = j + 1;
                int right = nums.Length - 1;
                while (left < right){
                    if (nums[left] + nums[right] == sum){
                        List<int> lst = new List<int>();
                        lst.Add(nums[i]); lst.Add(nums[j]);
                        lst.Add(nums[left]); lst.Add(nums[right]);
                        result.Add(lst);
                        left++; right--;
                        // Keep sure there's no duplicates
                        while (left < nums.Length && nums[left] == nums[left - 1]) left++;
                        while (right > j + 1 && nums[right] == nums[right + 1]) right--;
                    }
                    else if (nums[left] + nums[right] < sum) left++;
                    else if (nums[left] + nums[right] > sum) right--;
                }
            }
        }
        return result;
    }
}

File: 18_4sum.py
from typing import List


class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if len(nums) < 4: return []
        nums.sort()
        results = []
        resultsDict = {}
        for i in range(len(nums) - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            for j in range(i+1, len(nums) -2):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                left = j+1
                right = len(nums) - 1
                while left < right:
                    if nums[i] + nums[j] + nums[left] + nums[right] == target:
                        result = [nums[i], nums[j], nums[left], nums[right]]
                        result.sort()

                        resultKey = ''.join([str(x) for x in result])
                        if resultKey not in resultsDict:
                            results.append(result)
                            resultsDict[resultKey] =1
                    
                    if nums[i] + nums[j] + nums[left] + nums[right] < target:
                        left +=1
                    else:
                        right -=1

        
        return results


instance = Solution()
print(instance.fourSum([0,0,0,0],0))